### PR TITLE
feat: Introduce (incremental) Elo evaluation and use case for QA

### PR DESCRIPTION
**Pull request by SebastianNiehusAA**
# Description
- Implement generic and QA-specialized EloEvaluationLogic.
- Implement QA-specialized IncrementalEloQaEvaluationLogic.
- Add EloAggregationAdapter for compatibility with previous Argilla-centric Elo aggregation
- Add tutorial notebook describing the (incremental) Elo evaluation use case

## Before Merging
 - [x] Review the code changes
    - Unused print / comments / TODOs
    - Missing docstrings for functions that should have them
    - Consistent variable names
    - ...
 - [x] Update `changelog.md` if necessary
 - [x ] Commit messages should contain a semantic [label](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) and the ticket number
   - Consider squashing if this is not the case

### CODE REVIEW

**Comment by SebastianNiehusAA** [comment]
Note: Currently the IncrementalEloQaEvaluationLogic duplicates a lot of code from the EloQaEvaluationLogic and all of its functionality. We would therefore like to remove the EloQaEvaluationLogic.

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,13 +3,18 @@
 ## Unreleased
 
 ### Breaking Changes
-...
+ - Changed the behavior of `IncrementalEvaluator::do_evaluate`  such that it now promotes all output to `do_incremental_evaluate`instead of only the new outputs.
```
Comment (NiklasKoehneckeAA) [inline CHANGELOG.md:6]:
```suggestion
 - Changed the behavior of `IncrementalEvaluator::do_evaluate` such that it now sends all output to `do_incremental_evaluate` instead of only the new outputs.
```
Is output an actual `Output` ? otherwise its not quite clear what happens here

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,13 +3,18 @@
...
+ -
 ### New Features
-...
+ - Add generic `EloEvaluator` class and `EloEvaluationLogic`for implementation of Elo evaluation use cases.
```
Comment (NiklasKoehneckeAA) [inline CHANGELOG.md:9]:
```suggestion
 - Add generic `EloEvaluator` class and `EloEvaluationLogic` for implementation of Elo evaluation use cases.
```

File: tests/evaluation/test_elo_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+from intelligence_layer.connectors import AlephAlphaClientProtocol
+from intelligence_layer.core import (
+    ControlModel,
+    Language,
+    LuminousControlModel,
+    TextChunk,
+    utc_now,
+)
+from intelligence_layer.core.tracer.tracer import NoOpTracer, Tracer
```
Comment (NiklasKoehneckeAA) [inline tests/evaluation/test_elo_evaluator.py:14]: wrong imports

File: tests/evaluation/test_elo_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    InMemoryEvaluationRepository,
+    InMemoryRunRepository,
+    Matches,
+    MatchOutcome,
+    RunOverview,
+    SuccessfulExampleOutput,
+)
+from intelligence_layer.examples import SingleChunkQaInput, SingleChunkQaOutput
+
+load_dotenv()
```
Comment (NiklasKoehneckeAA) [inline tests/evaluation/test_elo_evaluator.py:32]: why is this here and not in some init?

File: tests/evaluation/test_elo_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+
+@fixture
+def in_memory_run_repository() -> InMemoryRunRepository:
+    return InMemoryRunRepository()
+
+
+@fixture
+def in_memory_evaluation_repository() -> InMemoryEvaluationRepository:
+    return InMemoryEvaluationRepository()
```
Comment (NiklasKoehneckeAA) [inline tests/evaluation/test_elo_evaluator.py:79]: I'm pretty sure these fixtures exist already in the top-level conftest

File: tests/evaluation/test_elo_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+                dataset_id=dataset_id,
+                id=run_ids[i],
+                start=utc_now(),
+                end=utc_now(),
+                failed_example_count=0,
+                successful_example_count=len(qa_outputs),
+                description="runner",
+            )
+        )
+    return run_ids, dataset_id
```
Comment (NiklasKoehneckeAA) [inline tests/evaluation/test_elo_evaluator.py:170]: this method is only used once and you don't use the dataset id

File: tests/evaluation/test_elo_evaluator.py
Comment (NiklasKoehneckeAA) [inline tests/evaluation/test_elo_evaluator.py:1]: why is this file called test_elo_evaluator when what it tests is a regular evaluator with elo logic

**Review by NiklasKoehneckeAA — COMMENTED** [review]
I'd like to see docstrings for public methods of classes that you expect people to use (which do not have docstrings yet on superclasses). The logics do not have this, which guess is ok if you think people will not really use them

File: CHANGELOG.md (hunk 1 above)
Comment (SebastianNiehusAA) [reply CHANGELOG.md:6]: Output is `SuccessfulExampleOutput`. Adapted the text accordingly.

File: tests/evaluation/test_elo_evaluator.py (hunk 5 above)
Comment (MerlinKallenbornAA) [reply tests/evaluation/test_elo_evaluator.py]: Removed them :)

File: tests/evaluation/test_elo_evaluator.py (hunk 6 above)
Comment (MerlinKallenbornAA) [reply tests/evaluation/test_elo_evaluator.py]: Good catch, removed `dataset_id`from the return

File: tests/evaluation/test_elo_evaluator.py
Comment (MerlinKallenbornAA) [reply tests/evaluation/test_elo_evaluator.py:1]: True, renamed it to `test_elo_evaluation_logic.py`

**Comment by MerlinKallenbornAA** [comment]
> I'd like to see docstrings for public methods of classes that you expect people to use (which do not have docstrings yet on superclasses). The logics do not have this, which guess is ok if you think people will not really use them


Yes good call! Added some docstrings
